compiler/rustc_lint/src/internal.rs

```diff
@@ -393,8 +393,14 @@ impl LateLintPass<'_> for Diagnostics {
             return;
         }
 
+        let mut found_parent_with_attr = false;
         let mut found_impl = false;
-        for (_, parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+        for (hir_id, parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+            if let Some(owner_did) = hir_id.as_owner() {
+                found_parent_with_attr |=
+                    cx.tcx.has_attr(owner_did.to_def_id(), sym::rustc_lint_diagnostics);
```

Comment:
I believe `|=` (BitOr) does not have short circuitry. So avoid doing so when attribute is already found.

```suggestion
                found_parent_with_attr = found_parent_with_attr ||
                    cx.tcx.has_attr(owner_did.to_def_id(), sym::rustc_lint_diagnostics);
```

Reply:
Fixed!